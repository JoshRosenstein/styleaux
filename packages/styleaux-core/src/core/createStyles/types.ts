import { Style, CSSObj } from '@styleaux/types';
import {
  CreateStylesValueGetterBoth,
  ExtractPrimitive,
  OmitTheme,
  Props,
  PropStyleFunc,
  PropStyleArrayFunc,
  WithTheme,
} from '../types';

type MaybeArr<T> = T | T[];

type StyleInputValueP<T, P> =
  | Style
  | CSSObj
  | MaybeArr<CreateStylesValueGetterBoth<T, P>>;

// Maybe shouldnt extract all primitives if prop value can be a nested object?
export type StyleInputFromProps<P> = OmitTheme<
  { [K in keyof P]?: StyleInputValueP<NonNullable<ExtractPrimitive<P[K]>>, P> }
>;

export interface CreateStyles<
  PROPS extends Props = Props,
  Theme extends {} = never,
  Media extends {} = never,
  PX = WithTheme<PROPS, Theme, Media>
> {
  (
    styles: StyleInputFromProps<PROPS>,
    staticOrStyleFunc?: StaticOrStyleFunc<PROPS, PX>,
  ): PropStyleArrayFunc<PX>;
}

export type StaticOrStyleFunc<P, PX> =
  | Style
  | CSSObj
  | PropStyleFunc<PX | P>
  | PropStyleArrayFunc<PX | P>;

export interface CreateCreateStyles<
  Theme extends {} = never,
  Media extends {} = never
> {
  <PROPS extends Props = Props, PX = WithTheme<PROPS, Theme, Media>>(
    styles: StyleInputFromProps<PROPS>,
    staticOrStyleFunc?: StaticOrStyleFunc<PROPS, PX>,
  ): PropStyleArrayFunc<PX>;
}
