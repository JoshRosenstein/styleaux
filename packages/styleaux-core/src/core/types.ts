import {
  DEFAULT_MEDIA_KEY,
  THEME_KEY,
  MEDIA_KEY,
  IConstants,
} from '../constants'
import {Media} from './overidableTypes'
import {Arg1,Omit,NonNever, Dictionary} from '../types'
import {Interpolation3} from '../cssTypes2'
export type MediaKey= string | number

export type PropStyleFunc<P extends Record<string,any>=Record<string,any>> = (mergedProps: P) => Interpolation3
export type PropStyleFuncArr<P extends Record<string,any>=Record<string,any>> = (mergedProps: P) => Interpolation3[]

export type GetStylePropsLazy<S,M> =  ResponsiveObject<EtractInputType<S>,M>


interface IAllCssValue<T> {
  /**
   * The base css value, without any media queries applied
   */
  [DEFAULT_MEDIA_KEY]?: T
}


type OmitNevers<M>=NonNever<M>

export type MediaMapFromKeys<Media extends {},V > = {
  [key in keyof OmitNevers<Media>]?: V
}  & IAllCssValue<V>

export type ResponsivePropValue<Media extends {}, ValueType> =
MediaMapFromKeys<Media,ValueType> | Array<ValueType | undefined>



type If<Condition extends boolean, Then, Else = never> = Condition extends true ? Then : Else;


export type IsNeverOrEmptyObj <T> =
[T] extends [never] ? true: keyof OmitNevers<T> extends never?true: false


export type ResponsiveProp<ValueType, BreakPoints = never> =ValueType | ResponsivePropValue<BreakPoints, ValueType>

export type NeverToStringOrNumber <T> =T extends never? string | number: T
export type NeverToStringOrNumberMap <T> = {[K in keyof T]?: NeverToStringOrNumber<T[K]>}


export type ResponsiveObject<P, B extends {}=Media> =
If<IsNeverOrEmptyObj<B>,NeverToStringOrNumberMap<P>,{
  [K in keyof P]?: ResponsiveProp<NeverToStringOrNumber<P[K]> , B>
}>


export type IfNever <T,Then> = [T] extends [never] ? Then: T



export type ResponsiveProps<P, B> = {
  [K in keyof P]?: ResponsiveProp< IfNever<P[K],string | number> , B>
}

export type WithOptionalTheme<P extends {[THEME_KEY]?: T}, T> = Omit<P, typeof THEME_KEY> & {
  [THEME_KEY]?: T
}

export type ThemeWithOptionalMedia<T extends {[MEDIA_KEY]?: M}, M> = Omit<T, typeof MEDIA_KEY> & {
  [MEDIA_KEY]?: M
}

export interface ITheme<T> {
  [THEME_KEY]?: T & {[MEDIA_KEY]?: Media}
}
export interface IBreakpointTheme<T, B> {
  [THEME_KEY]?: T & {[MEDIA_KEY]: B}
}


export type ThemeWithBreakpoints<T, B> = [B] extends [never]
  ? [T] extends [never]
    ? Partial<ITheme<any>>
    : ITheme<T>
  : [T] extends [never]?{[THEME_KEY]?:any}: IBreakpointTheme<T, B>

export type WithTheme<P, T, B> = ResponsiveObject<P, B> &
  ThemeWithBreakpoints<T, B>

export type ExtractNonResponsiveValue<T> = T extends ResponsiveProp<
  infer U,
  any
>
  ? U
  : T

  export type InferResponsivePropValue<T> = T extends ResponsiveProp<
  infer U,
  any
>
  ? U
  : T

  export interface IStyles {
    [ruleOrSelector: string]: string | number | IStyles;
  }

export type Indices<T> = Exclude<keyof T, keyof any[]>
export type ExtractArg1FromArray<T> = Arg1<T[Indices<T>]>

export type OmitTheme<T extends {},K=IConstants['THEME_KEY']>= K extends keyof T? Omit<T,K>:T

export type OmitStaticStyles<R,K= IConstants['STATIC_STYLES_KEY']> = K extends keyof R
  ? Omit<R,K>:R

  export type ResolveCreateStyleType<T>=T extends IStyles
  ? boolean
  : T extends (...args: any[]) => any
  ? Arg1<T>
  : T extends ((...args: any[]) => any)[]
  ? ExtractArg1FromArray<T>
  : T extends Dictionary
  ? boolean
  // If primitive, then prop types was directly passed
  : T extends number | string | boolean
  ? T : {}

export type EtractInputType<P> = OmitStaticStyles<
  {
    [K in keyof P]: ResolveCreateStyleType<P[K]>
  }
>

export type EtractNonResponsiveInputType<STYLES, I = EtractInputType<STYLES>> = {
  [K in keyof I]?: InferResponsivePropValue<I[K]>
}


export type InferPropsFromFunctionArgument<T> = T extends (
  props: infer R,
) => any
  ? R //OmitTheme<R>
  : {};

