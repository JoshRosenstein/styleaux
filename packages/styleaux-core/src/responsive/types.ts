import {Dict, Maybe} from '../types'

export type BreakPointKeysOFT<B, T> = B extends Dict<any>
  ? Partial<Record<keyof B | 'default', T>> | (T)[]
  : B extends any[]
  ? (T)[]
  : never

export type BreakPointKeysOfNumberOrString<B> = BreakPointKeysOFT<
  B,
  number | string
>

export type MaybeInputNumberOrString<B> = Maybe<
  BreakPointKeysOfNumberOrString<B> | string | number
>

export type IBreakpoints = Dict<number | string> | (number | string)[]

export type ToMq = (x: any) => string

export type TransformStyle = (
  {value}: {[index: string]: any; value: any},
) => (props: any) => any

export type InputMaybe<B> = Maybe<
  BreakPointKeysOfNumberOrString<B> | string | number
>

// type ValueTypes<T> = T extends Dict<number | string>
//   ? Partial<Record<keyof T | 'default', number | string>> | (number | string)[]
//   : never
