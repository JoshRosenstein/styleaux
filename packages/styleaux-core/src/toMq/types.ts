import {IDictionary} from '../types'

export type OptionsT<M extends PropertyKey> = {
  toMq?: {valueConverter?: Function | M}
}

export interface toMqInputAsObj extends IDictionary {
  min?: string | number
  max?: string | number
  minW?: string | number
  maxW?: string | number
  minH?: string | number
  maxH?: string | number
  'max-width'?: string | number
  'max-height'?: string | number
  'min-width'?: string | number
  'min-height'?: string | number
  maxWidth?: string | number
  maxHeight?: string | number
  minWidth?: string | number
  minHeight?: string | number
  handheld?: boolean
  screen?: boolean
  print?: boolean
  all?: boolean
  orientation?: 'landscape' | 'portrait'
}

// type UnionKeyToValue<U extends string, T> = {[K in U]: T}
// type KnownKeys<T> = {
//   [K in keyof T]: string extends K ? never : number extends K ? never : K
// } extends {[_ in keyof T]: infer U}
//   ? U
//   : never
