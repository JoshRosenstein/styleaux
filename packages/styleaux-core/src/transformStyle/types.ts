export interface ILocalOptions<P> {
  getterFn?: Function
  path?: string
  postFn?: any
  preFn?: any
  props?: P
  defaultTransform?: boolean
  defaultLookup?: boolean
  //key?: string /// OLD
  //path?: string
  // getter?: transformFuncs /// OLD
  // postFn?: transformFuncs
  //preFn?: transformFuncs
}

export interface transformStyleOptions<P> extends ILocalOptions<P> {
  value: string | number
  cssProp?: string
  valueOnly?: boolean
}

export type transformStyleConfig = {
  /**
   * TODO
   */
  value: string | number
  /**
   * TODO
   */
  cssProp?: string
  /**
   * TODO
   */
  valueOnly?: boolean
  /**
   * TODO
   */
  path?: string
  /**
   * TODO
   */
  postFn?: any
  /**
   * TODO
   */
  preFn?: any
}

export type transformStylePConfig = {
  /**
   * TODO
   */
  value: string | number
  /**
   * TODO
   */
  cssProp?: string
  /**
   * TODO
   */
  valueOnly?: boolean
  /**
   * TODO
   */
  path?: string
  /**
   * TODO
   */
  postFn?: any
  /**
   * TODO
   */
  preFn?: any
}
