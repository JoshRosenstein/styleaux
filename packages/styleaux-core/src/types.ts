export type Nothing = null | undefined
export type Maybe<T> = Nothing | T

export type Primitive = string | number | boolean | undefined | null
export type Dict<T, K extends string | number = string> = {[key in K]: T}

export type AnyDict = Dict<any>
export type MaybeAnyDict = Maybe<AnyDict>

export type EmptyDict = Dict<never>

export type DictValues<T> = T extends Dict<infer U> ? U : never
export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends Array<infer U>
    ? Array<DeepPartial<U>>
    : T[P] extends ReadonlyArray<infer U>
    ? ReadonlyArray<DeepPartial<U>>
    : DeepPartial<T[P]>
}

export type MapKeys<T> = {[K in keyof T]: T[K]}
