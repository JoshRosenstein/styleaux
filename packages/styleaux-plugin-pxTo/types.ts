// @flow

export type pxToStrT = (pxValue: number | string) => string
export type pxToNumT = (pxValue: number | string) => number
export type pxToT = {
  <T extends number>(unit: undefined): (pxValue: T) => T
  <T extends string>(unit: T): (pxValue: number | T) => T

  (unit: string | undefined): (
    pxValue: number | string | void,
  ) => number | string | void
}
