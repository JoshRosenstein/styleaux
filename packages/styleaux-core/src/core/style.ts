import { identity } from "@roseys/futils";
import { CSSPropertyKeys } from "../types";
import { isNumber, isBoolean, isString, isFunction } from "typed-is";
import { everyMedia } from "./everyMedia";
import { createWrap } from "../utils/wrap";
import {ResponsiveProp} from './types'

export type KeysOrstrong<T extends {}=never>=T extends [never] ? string : Extract<keyof T,string>

export function style<T>({
  cssProp,
  getStyle = createWrap(cssProp),
  getValue = identity
}: {
  cssProp?: CSSPropertyKeys;
  getStyle?: (result: any, input?: any, props?: any, mediaKey?: string) => any;
  getValue?: (input: any, props?: any, mediaKey?: string) => any;
}) {
  function getValues<Media1 extends {}=never>(
    get: (
      input: any,
      props?: {},
      mediaKey?: KeysOrstrong<Media1>
    ) => boolean | Function | null | undefined | string | number,
    input,
    props,
    mediaKey
  ) {
    const result = get(input, props, mediaKey);

    if (isBoolean(result)) {
      return null;
    }

    if (result === undefined && (isString(input) || isNumber(input))) {
      return input;
    }
    return isFunction(result)
      ? getValues(result, input, props, mediaKey)
      : result;
  }

  return <Media extends {}=never,M extends string =KeysOrstrong<Media>>(input: ResponsiveProp<T,Media>, props: {}, mediaKey?: M) => {
    return everyMedia(
      props,
      getValues(getValue, input, props, mediaKey),
      result => getStyle(result, input, props, mediaKey)
    );
  };
}
