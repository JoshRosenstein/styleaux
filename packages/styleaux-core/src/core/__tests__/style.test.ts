import { WithTheme } from '../types';
import { Arg1, AnyFunc } from '../../types';
import { style, extendStyle } from '../style';
import { prop, merge, pick } from '@roseys/futils';
type Media = { sm: string; md: string };
const media: Media = {
  sm: '40em',
  md: '52em',
};

const theme = {
  media,
  space: [0, '10px', '20px'],
  colors: {
    blue: '#07c',
    black: '#111',
  },
};
type Theme = typeof theme;
const themeArr = {
  media: ['40em', '52em'],
  colors: {
    blue: '#07c',
    black: '#111',
  },
};
type ThemeArr = typeof themeArr;

export function snapshot<Fn extends AnyFunc>(fn: Fn, pickthemeKeys?: string[]) {
  return (props: Arg1<Fn>, label?: string) => {
    const NewProps = props.theme
      ? merge(props, {
          theme: pick(
            pickthemeKeys === undefined
              ? ['media']
              : pickthemeKeys.concat('media'),
            prop('theme', props),
          ),
        })
      : props;

    let snapshotName = JSON.stringify(NewProps, null, 2);
    if (label) {
      snapshotName = `[${label}] ${snapshotName}`;
    }
    expect(fn(props)).toMatchSnapshot(snapshotName);
  };
}

const width = style<
  WithTheme<{ width: number | string }, Theme | ThemeArr, Media>
>({
  prop: 'width',
});
const testWidth = snapshot(width);
const widthArr = style<
  WithTheme<{ width: number | string }, ThemeArr, string[]>
>({
  prop: 'width',
});
const testWidthArr = snapshot(widthArr);
const backgroundColor = style<
  WithTheme<{ backgroundColor: string; bg: string }, Theme | ThemeArr, Media>
>({
  prop: 'backgroundColor',
  alias: 'bg',
  key: 'colors',
});

const testBg = snapshot(backgroundColor, ['colors']);

const margin = style<{ margin: string | number }>({
  prop: 'margin',
  cssProp: 'margin',
  transformValue: (n) => n + 'px',
  key: 'space',
});
const testMargin = snapshot(margin, ['space']);

const testmarginY = snapshot(
  style<{ marginY: string | number; my: string | number }, Theme, Media>({
    cssProp: ['marginTop', 'marginBottom'],
    prop: ['marginY', 'my'],
    key: 'space',
  }),
  ['space'],
);

describe('Style', () => {
  test('returns values from theme', () => {
    const styleFn = style<
      WithTheme<{ color: string }, Theme | ThemeArr, Media>
    >({
      prop: 'color',
      key: 'colors',
    });
    const testProps = snapshot(styleFn, ['colors']);
    testProps({ theme, color: 'blue' });
  });

  test('handles aliased props', () => {
    testBg({
      theme,
      bg: 'blue',
    });
  });

  test('returns 0', () => {
    testWidth({ width: 0 });
  });

  test('returns responsive style objects', () => {
    testWidth({
      theme,
      width: { all: '100%', sm: '50%' },
    });
  });

  test('returns responsive style objects from array with media as array', () => {
    testWidthArr({
      theme: themeArr,
      width: ['100%', '50%'],
    });
  });

  test('returns responsive style objects from array with media as object', () => {
    testWidth({
      theme,
      width: ['100%', '50%'],
    });
  });

  test('skips undefined responsive values', () => {
    testWidth({
      theme,
      width: ['100%', , '50%'],
    });
  });

  test('looksup from array Theme', () => {
    testMargin({ margin: 1, theme: { space: [0, 11, 22, 33, 44] } });
  });

  test('looksup from array Theme with Negative', () => {
    testMargin({ margin: -1, theme: { space: [0, 11, 22, 33, 44] } });
  });

  test('looksup from Object Theme', () => {
    testMargin({ margin: 'sm', theme: { space: { sm: 11 } } });
  });

  test('looksup from Object Theme with negative', () => {
    testMargin({ margin: '-sm', theme: { space: { sm: 11 } } });
  });

  test('uses input if not themeKey', () => {
    testMargin({ margin: 20 });
  });

  test('Can Handle Multiple PropKeys', () => {
    const testcolor = snapshot(
      style<{ color: string; fg: string }>({
        cssProp: 'color',
        prop: ['color', 'fg'],
        key: 'colors',
      }),
      ['colors'],
    );
    testcolor({ fg: 'blue' }, 'fg');
    testcolor({ color: 'black' }, 'color');
    testcolor({ color: 'black', fg: 'blue' }, 'First Prop');
  });

  test('Can Handle Multiple PropKeys & cssKeys', () => {
    testmarginY({ marginY: '1px' }, 'marginY');
    testmarginY({ marginY: '1px' }, 'my');
  });
  test('Can Handle Multiple PropKeys & cssKeys Responsive', () => {
    testmarginY({ marginY: { sm: 1, md: 2 }, theme }, 'marginY');
    testmarginY({ my: { sm: 1, md: 2 }, theme }, 'my');
  });
});

describe('Extends', () => {
  test('returns style object array', () => {
    const styleFunc = extendStyle({})({ prop: ['a', 'b'] });
    const testProps = snapshot(styleFunc);
    testProps({ a: 'foo' });
    testProps({ b: 'bar' });
  });

  test('overrides base options', () => {
    const styleFact = extendStyle({
      prop: ['a', 'b'],
      cssProp: ['x', 'y'],
    });
    const styleFunc = styleFact({
      prop: ['c', 'd'],
    });
    const testProps = snapshot(styleFunc);
    testProps({ c: 'C' });
    testProps({ d: 'D' });
  });
});
