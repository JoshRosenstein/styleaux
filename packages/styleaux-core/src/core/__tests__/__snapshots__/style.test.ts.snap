// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extends overrides base options: {
  "c": "C"
} 1`] = `
Array [
  Object {
    "x": "C",
    "y": "C",
  },
]
`;

exports[`Extends overrides base options: {
  "d": "D"
} 1`] = `
Array [
  Object {
    "x": "D",
    "y": "D",
  },
]
`;

exports[`Extends returns style object array: {
  "a": "foo"
} 1`] = `
Array [
  Object {
    "a": "foo",
    "b": "foo",
  },
]
`;

exports[`Extends returns style object array: {
  "b": "bar"
} 1`] = `
Array [
  Object {
    "a": "bar",
    "b": "bar",
  },
]
`;

exports[`Style Can Handle Multiple PropKeys & cssKeys Responsive: [marginY] {
  "marginY": {
    "sm": 1,
    "md": 2
  },
  "theme": {
    "space": [
      0,
      "10px",
      "20px"
    ],
    "media": {
      "sm": "40em",
      "md": "52em"
    }
  }
} 1`] = `
Array [
  Object {
    "@media screen and (min-width: 40em)": Object {
      "marginBottom": "10px",
      "marginTop": "10px",
    },
    "@media screen and (min-width: 52em)": Object {
      "marginBottom": "20px",
      "marginTop": "20px",
    },
  },
]
`;

exports[`Style Can Handle Multiple PropKeys & cssKeys Responsive: [my] {
  "my": {
    "sm": 1,
    "md": 2
  },
  "theme": {
    "space": [
      0,
      "10px",
      "20px"
    ],
    "media": {
      "sm": "40em",
      "md": "52em"
    }
  }
} 1`] = `
Array [
  Object {
    "@media screen and (min-width: 40em)": Object {
      "marginBottom": "10px",
      "marginTop": "10px",
    },
    "@media screen and (min-width: 52em)": Object {
      "marginBottom": "20px",
      "marginTop": "20px",
    },
  },
]
`;

exports[`Style Can Handle Multiple PropKeys & cssKeys: [marginY] {
  "marginY": "1px"
} 1`] = `
Array [
  Object {
    "marginBottom": "1px",
    "marginTop": "1px",
  },
]
`;

exports[`Style Can Handle Multiple PropKeys & cssKeys: [my] {
  "marginY": "1px"
} 1`] = `
Array [
  Object {
    "marginBottom": "1px",
    "marginTop": "1px",
  },
]
`;

exports[`Style Can Handle Multiple PropKeys: [First Prop] {
  "color": "black",
  "fg": "blue"
} 1`] = `
Array [
  Object {
    "color": "black",
  },
]
`;

exports[`Style Can Handle Multiple PropKeys: [color] {
  "color": "black"
} 1`] = `
Array [
  Object {
    "color": "black",
  },
]
`;

exports[`Style Can Handle Multiple PropKeys: [fg] {
  "fg": "blue"
} 1`] = `
Array [
  Object {
    "color": "blue",
  },
]
`;

exports[`Style handles aliased props: {
  "theme": {
    "colors": {
      "blue": "#07c",
      "black": "#111"
    },
    "media": {
      "sm": "40em",
      "md": "52em"
    }
  },
  "bg": "blue"
} 1`] = `
Array [
  Object {
    "backgroundColor": "#07c",
  },
]
`;

exports[`Style looksup from Object Theme with negative: {
  "margin": "-sm",
  "theme": {
    "space": {
      "sm": 11
    }
  }
} 1`] = `
Array [
  Object {
    "margin": "-11px",
  },
]
`;

exports[`Style looksup from Object Theme: {
  "margin": "sm",
  "theme": {
    "space": {
      "sm": 11
    }
  }
} 1`] = `
Array [
  Object {
    "margin": "11px",
  },
]
`;

exports[`Style looksup from array Theme with Negative: {
  "margin": -1,
  "theme": {
    "space": [
      0,
      11,
      22,
      33,
      44
    ]
  }
} 1`] = `
Array [
  Object {
    "margin": "-11px",
  },
]
`;

exports[`Style looksup from array Theme: {
  "margin": 1,
  "theme": {
    "space": [
      0,
      11,
      22,
      33,
      44
    ]
  }
} 1`] = `
Array [
  Object {
    "margin": "11px",
  },
]
`;

exports[`Style returns 0: {
  "width": 0
} 1`] = `
Array [
  Object {
    "width": 0,
  },
]
`;

exports[`Style returns responsive style objects from array with media as array: {
  "theme": {
    "media": [
      "40em",
      "52em"
    ]
  },
  "width": [
    "100%",
    "50%"
  ]
} 1`] = `
Array [
  Array [
    Object {
      "width": "100%",
    },
    Object {
      "@media screen and (min-width: 40em)": Object {
        "width": "50%",
      },
    },
  ],
]
`;

exports[`Style returns responsive style objects from array with media as object: {
  "theme": {
    "media": {
      "sm": "40em",
      "md": "52em"
    }
  },
  "width": [
    "100%",
    "50%"
  ]
} 1`] = `
Array [
  Array [
    Object {
      "width": "100%",
    },
    Object {
      "@media screen and (min-width: 40em)": Object {
        "width": "50%",
      },
    },
  ],
]
`;

exports[`Style returns responsive style objects: {
  "theme": {
    "media": {
      "sm": "40em",
      "md": "52em"
    }
  },
  "width": {
    "all": "100%",
    "sm": "50%"
  }
} 1`] = `
Array [
  Object {
    "@media screen and (min-width: 40em)": Object {
      "width": "50%",
    },
    "width": "100%",
  },
]
`;

exports[`Style returns values from theme: {
  "theme": {
    "colors": {
      "blue": "#07c",
      "black": "#111"
    },
    "media": {
      "sm": "40em",
      "md": "52em"
    }
  },
  "color": "blue"
} 1`] = `
Array [
  Object {
    "color": "#07c",
  },
]
`;

exports[`Style skips undefined responsive values: {
  "theme": {
    "media": {
      "sm": "40em",
      "md": "52em"
    }
  },
  "width": [
    "100%",
    null,
    "50%"
  ]
} 1`] = `
Array [
  Array [
    Object {
      "width": "100%",
    },
    undefined,
    Object {
      "@media screen and (min-width: 52em)": Object {
        "width": "50%",
      },
    },
  ],
]
`;

exports[`Style uses input if not themeKey: {
  "margin": 20
} 1`] = `
Array [
  Object {
    "margin": "20px",
  },
]
`;
